api/metric/api.go

```diff
@@ -18,64 +18,169 @@ import (
 	"context"
 
 	"go.opentelemetry.io/api/core"
-	"go.opentelemetry.io/api/registry"
 	"go.opentelemetry.io/api/unit"
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
+//go:generate stringer -type=Kind
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid        Kind = iota
+	CumulativeKind      // Supports Add()
+	GaugeKind           // Supports Set()
+	MeasureKind         // Supports Record()
 )
 
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Recorder interface {
+	// Record allows the SDK to observe a single metric event
+	Record(ctx context.Context, value float64)
+}
+
+// LabelSet represents a []core.KeyValue for use as pre-defined labels
+// in the metrics API.
+//
+// TODO this belongs outside the metrics API, in some sense, but that
+// might create a dependency.  Putting this here means we can't re-use
+// a LabelSet between metrics and tracing, even when they are the same
+// SDK.
+type LabelSet interface {
+	Meter() Meter
+}
+
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
 type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+	// DefineLabels returns a reference to a set of labels that
+	// cannot be read by the application.
+	DefineLabels(context.Context, ...core.KeyValue) LabelSet
+
+	// RecorderFor returns a handle for observing single measurements.
+	RecorderFor(context.Context, LabelSet, Instrument) Recorder
+
+	// RecordSingle records a single measurement without computing a handle.
+	RecordSingle(context.Context, LabelSet, Measurement)
+
+	// RecordBatch atomically records a batch of measurements.  An
+	// implementation may elect to call `RecordSingle` on each
+	// measurement, or it could choose a more-optimized approach.
+	RecordBatch(context.Context, LabelSet, ...Measurement)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+type InstrumentID uint64
+
+// Instrument represents a named metric with recommended local-aggregation keys.
+type Instrument struct {
+	// Name is a required field describing this metric instrument,
+	// should have length > 0.
+	Name string
+
+	// ID is uniquely assigned to support per-SDK registration.
+	ID InstrumentID
+
+	// Description is an optional field describing this metric instrument.
+	Description string
+
+	// Unit is an optional field describing this metric instrument.
+	Unit unit.Unit
+
+	// Kind is the metric kind of this instrument.
+	Kind Kind
+
+	// Bidirectional implies this is an up-down Cumulative.
```

Comment:
What is the use case for know if an Instrument could be Bidirectional and Unidirectional? Do we need both variables or should this be changed to a single variable? if I get it correctly, maybe it could be renamed to `Monotonic`

Reply:
This is a great point, and it belongs in the RFC 0003 discussion.. Please use this PR: https://github.com/open-telemetry/oteps/pull/51

I like the idea of replacing "Unidirectional" with "Monotonic": this option applies to Gauges and is *not the default* behavior.

For Counters, Monotonic *is the default* behavior, so we still need a name for the option to support bi-directional counters. I'm not aware of a good opposite for monotonic. Do you have any suggestions?

Reply:
Sent the discussion to mentioned PR.

Reply:
@krnowak I suppose the `Option` objects could be made specific to the particular constructor, so that `NewCounter(..., NonMononic())` would build but `NewCounter(..., NonNegative)` would not.

Also FYI I updated the RFCs to use "Monotonic" but we're still working out whether "NonMonotonic" is a preferred to "BiDirectional".